include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,519 @@
+ *     </td>
+ *   </tr>
+ * </table>
+ *
+ * <h3>Implementation</h3>
+ *
+ * The implementation of the class is based on FESystem which is aggregated as
+ * a private member and initialized for the case above as
+ *
+ * @code
+ * FESystem<dim> fe_system(FE_Q<dim>(1),1,
+ *                         FE_Q<dim>(1),1);
+ * @endcode
+ *
+ * Note that it would not be wise to have this class derived
+ * from FESystem as the latter is a vector-valued FE, whereas the current class
+ * can be both scalar and vector valued depending on the base FEs used to build it.
+ *
+ * The ordering of the shape function, @p interface_constrains, the @p prolongation (embedding)
+ * and the @p restriction matrices are taken from the FESystem class.
+ *
+ * <h3>hp-constraints</h3>
+ *
+ * This class implements hp constraints with arbitrary FE. For the example above,
+ * a user will need to wrap a given FiniteElement into FE_Enriched:
+ * @code
+ *   FE_Enriched<dim> fe_enriched(FE_Q<dim>(1));
+ * @endcode
+ *
+ * This corresponds to building constraints between enriched FE and
+ *
+ * @code
+ * FESystem<dim>(FE_Q<dim>(1),1,
+ *               FE_Nothing<dim>(1,true),1)
+ * @endcode
+ *
+ * In order to make the resulting field continuous
+ * all enriched DoFs on the interface are constrained to be zero.
+ * This is a natural outcome of FE_Nothing with domination set to <code>true</code>.
+ *
+ * Note, that if the constraints are built between two different instances of this class,
+ * no checks are performed to ensure that the underlying enrichment function is
+ * the same. Thus the user must be careful in order not to end up with the
+ * discontinuous FE space.
+ *
+ * When using this class, please cite
+ * @code{.bib}
+ @Article{Davydov2016,
+    author =    {Denis Davydov and Tymofiy Gerasimov and Jean-Paul Pelteret and Paul Steinmann},
+    title =     {On the h-adaptive PUM and hp-adaptive FEM approaches applied to PDEs in quantum mechanics.},
+    journal =   {Journal of Computational Physics. Submitted},
+    year =     {2016},
+ }
+ * @endcode
+ *
+ * @ingroup fe
+ *
+ * @author Denis Davydov, 2016.
+ */
+template <int dim, int spacedim=dim>
+class FE_Enriched : public FiniteElement<dim, spacedim>
+{
+public:
+
+  /**
+   * Constructor which takes base FiniteElement @p fe_base and the enrichment
+   * FiniteElement @p fe_enriched which will be multiplied by the @p enrichment_function.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base,
+               const FiniteElement<dim,spacedim> &fe_enriched,
+               const Function<dim>      *enrichment_function);
+
+  /**
+   * Constructor which only wraps the base FE @p fe_base.
+   * As for enriched FiniteElement, FE_Nothing is used.
+   * Continuity constraints will be automatically generated when
+   * this non-enriched element is used in conjunction with enriched finite element
+   * within the hp::DoFHandler.
+   */
+  FE_Enriched (const FiniteElement<dim,spacedim> &fe_base);
+
+
```

Comment:
remove one empty line, we typically only put one empty line between function declarations inside a class.